### DISCUSSION TITLE
How to properly document the scripts, aliases and practice we use for a specific robot setup

### DISCUSSION THREAD

**Discussion by S-Dafarra** [robotology/community#489] [Feedback & Meta]
When working on a specific robot setup, we often end up writing some scripts or aliases to ease its usage. 
Just to give an example, for ``iCubGenova04`` and ``iCubGenova09`` we use some alias to quickly go to the corresponding ``robots-configuration`` folder and to install the files, or to connect to the Bluetooth joystick.

While these may help the users that use these setups daily, it might not be trivial for people in the support team.

I wonder if there could be a commonly known location to store some "How To"s, "Tips and tricks" or similar.

cc @pattacini @vtikha @DanielePucci @traversaro @maggia80 @violadelbono @kouroshD @GiulioRomualdi @prashanthr05

**Comment by pattacini** [comment]
Hi @S-Dafarra 

Don't you have some wiki/webpages you have been maintaining within your group (maybe on https://github.com/dic-iit) that could serve to publish such information?

If yes, the support team can consult that documentation when in need.

**Reply by traversaro** [reply]
I would add that any robot-specific documentation could be linked in the robot descriptions in:
* https://robot-bazaar.iit.it/robots/collection/robots/iCubGenova04
* https://robot-bazaar.iit.it/robots/collection/robots/iCubGenova09

Note that the descriptions of the robots can be modified by any users by just clicking on the "Edit description button" by any user, and a PR will be automatically created to update the description.

**Reply by S-Dafarra** [reply]
Yes, I suppose there are a bunch of places in our organization as well where we could add such documentation. But then, the problem would be for a naive user to find that documentation. I was looking for a sort of default getaway where to properly add or link the relevant pages, independently from the specific research group. 
For example, ``icub-tech-support`` is the goto for any problem with the robot. I was wondering what would be the equivalent to document how to use a specific robot. Some important pieces of information would be
- account and passwords
- laptops involved
- location of the code
- use of shared directories or not
- location of the configuration files
- yarp network setup
- customized yarpmanager applications 
- scripts and aliases
- ....

**Reply by pattacini** [reply]
>Just to give an example, for iCubGenova04 and iCubGenova09 we use some alias

This was a bit misleading as I thought you were referring to scripts targeting a specific robot and customized for a specific group.

I'm in favor, for example, of collecting information into https://icub-tech-iit.github.io/documentation not specifically tailored for a single iCub system. We've already started doing that, actually.

Other than that, every single group should be responsible for publishing/maintaining their own documentation.

**Reply by S-Dafarra** [reply]
> This was a bit misleading as I thought you were referring to scripts targeting a specific robot and customized for a specific group.

I am indeed referring to scripts targeting a specific robot :grin: 

> Other than that, every single group should be responsible for publishing/maintaining their own documentation.

and I agree with that. But suppose I write the documentation on ``my_fancy_website.com``, how can I let people outside my group how to find it?

It is true that a group is responsible to maintain his robot, but these are also shared resources. So, I was wondering what could be a good place to put relevant information regarding how that system is setup.

Let me make another example. With ``iCubGenova04``  we decided to not use the ``.local`` installation for the robot configuration files because of a set of reasons. We are then usually modifying the source files and installing them. 
It also usually happens that we break the robot, and the fix requires a modification to the configuration files. Then, it often happens that the person that needs to do the fix does not know where to locate these files, and it has already happened that the installed files were modified instead. A few days later the fix was lost and we broke the robot again.

Now, we could write a detailed documentation somewhere on Github, maybe in a private repo too, but it is pretty likely that the person who is going to do the fix, cannot know where this documentation is located either, because this knowledge is limited to the group that wrote it.

Hence, the goal of this discussion was to break this last obstacle, making the setup documentation easily findable also outside the group that wrote it.

**Reply by S-Dafarra** [reply]
Let me also add that I am totally ok with the solution proposed by @traversaro in https://github.com/robotology/community/discussions/489#discussioncomment-433175 if we decide that that is the way to go

**Reply by pattacini** [reply]
>Let me also add that I am totally ok with the solution proposed by @traversaro in #489 (reply in thread) if we decide that that is the way to go

Indeed, the Bazaar is also a place where to collect information on groups working with our robots and therefore it is definitely appropriate advertizing specific resources including the examples you mentioned.

**Reply by pattacini** [reply]
Let me tag @vtikha who's responsible for the Bazaar.

**Reply by traversaro** [reply]
Yes, I think that if the support team needs to understand the state of the robot, the Bazaar page is the first place in which they need to look, even just for links to robot specific resources.

**Reply by S-Dafarra** [reply]
With @giovannipizzolante we tried to edit the Bazaar page, but none of the sections seemed a good fit. I guess the sections are "rigid" right?

**Reply by pattacini** [reply]
@S-Dafarra to continue discussing the use of Bazaar, maybe [icub-tech-iit/appsAway](https://github.com/icub-tech-iit/appsAway) turns to be a better place.

**Reply by pattacini** [reply]
You may want to link back here the issue then.

**Reply by vtikha** [reply]
Yes, please go ahead in `appsAway` and we can discuss to create a section that can fit the above requirements such as the one done for the `configuration-reports`. Maybe even a generic documentation entry. 

Sorry for the late reply

**Reply by vtikha** [reply]
During our discussions and various interactions with @S-Dafarra @giovannipizzolante, we decided to go ahead with the use of `custom fields` that can be added by the user. The user can add multiple fields as one would in a wiki or an issue composed in markdown. These customizable fields appear at the end of all the other sections, while in editing mode the customization appears as a new section which includes the possibility to add, modify or delete these newly added fields.

Here is a simple example to show the outcome/formatting:

<img width="575" alt="Screenshot 2021-05-05 at 16 10 59" src="https://user-images.githubusercontent.com/4537987/117156296-cfb37180-adbd-11eb-8a8d-3b3e14aaf58d.png">

**Reply by pattacini** [reply]
Looks awesome!
